0.4.2

rougetimelord opened the pull request:
This is a long-lived branch, intended to aggregate all changes planned for version `0.4.2`.

Pull requests should be made against this branch. Merges to this branch should be done via a merge commit. When ready, this branch should be merged to main via a squash commit.

## Changelog
- [chore] bump version string
- [fix] change condition for checking legacy verified DB if user used blue features
  - PR: #266
- [doc] add all properties to the BlueBlockerUser type definition
  - PR: #268 
- [fix] refix skipping deleted tweets
  - PR: #269
- [fix] remove reference to chrome.runtime.restartAfterDelay
  - Removes a firefox extension API incompatibility
  - PR: #271
- [feat] add a way for external extensions to register themselves as an integration
  - PR: #274 
- [chore] change firefox release workflow
    - Please work this time! 😭
    - H/T: @kheina, PR: #277 
- [feat] add blocking users for words/characters in their display name
  - H/T @MaxIsJoe, PR: #275
  - Closes #114 
- [feat] add ability to skip blocking for blue users
  - H/T @MaxIsJoe, PR: #278 
  - Closes #166 
- [fix] allow extension to run on on x.com
  - H/T @kheina, PR: #282
  - Closes #280  
- [chore] run prettier against all code

## Deployment Checklist
1. [x] merge all pull requests to llb
2. [x] ensure `src/manifest.ts` and `package.json` have the correct version number
3. [ ] use makefile to generate zips (`make chrome`, `make firefox`)
    - [x] chrome should be tested with `npm run build`
    - [ ] test firefox locally using zip
4. [ ] merge llb to main

kheina commented on src/shared.ts:
```diff
@@ -645,16 +646,31 @@ export async function BlockBlueVerified(user: BlueBlockerUser, config: Config) {
+		if (disallowedWordsWithoutEmptyStrings.length > 0){
+			// this makes extra sure that emojis are always detected, regardless if they are attached to a word or another string of emojis.
+			// the 'i' makes the test case insensitive, which helps users not have to worry about typing the same word multiple times with different variations.
+			const disallowedWordsAndEmojis = new RegExp(`${config.disallowedWords.map(word=>word.split('').join(' *')).join('(?= |$)|')}(?= |$)`, 'i');
+			const usernameToTest = (user.legacy.name).replace(/(?<= ) +/, '');
```
I am having a really hard time discerning why all of this is necessary

```js
>> `${["one", "two", "three"].map(word=>word.split('').join(' *')).join('(?= |$)|')}(?= |$)`
"o *n *e(?= |$)|t *w *o(?= |$)|t *h *r *e *e(?= |$)" 
```

this actually seems like the opposite of what we'd want?

I'm also not sure why it's necessary to modify the username. it looks like you're trying to replace multiple spaces with a single space but in a very roundabout way, `replace(/ {2,}/, ' ')` would be much more effective, but I still don't know why it's necessary?

rougetimelord replied:
In terms of the adding a space between characters part, I was mostly hypothesizing that some Twitter users might try to do "B A D W O R D", as a stylistic choice but I'm not sure how realistic that is?

For the editing usernames, same kinda deal. Your regex is a lot better, I honestly was kinda lost in the sauce of lookahead/behind groups lmao.

kheina commented on src/background/index.ts:
```diff
@@ -38,102 +68,199 @@ api.storage.sync.onChanged.addListener(async items => {
 ////////////////////////////////////////////////// EXTERNAL MESSAGE HANDLING //////////////////////////////////////////////////
 
-const [blockActionV1, blockAction] = ["BLOCK", "block_user"];
+const [blockActionV1, blockAction] = ['BLOCK', 'block_user', 'register'];
```
```suggestion
const [blockActionV1, blockAction, registerAction] = ['BLOCK', 'block_user', 'register'];
```

kheina commented on src/background/index.ts:
```diff
@@ -38,102 +68,199 @@ api.storage.sync.onChanged.addListener(async items => {
+		).integrations as { [id: string]: Integration };
+		const senderId = sender.id ?? '';
+		if (!integrations.hasOwnProperty(senderId)) {
+			if (message?.action == 'register') {
```
```suggestion
			if (message?.action === registerAction) {
```

kheina commented on src/background/index.ts:
```diff
@@ -38,102 +68,199 @@ api.storage.sync.onChanged.addListener(async items => {
+			await api.storage.local.get({ soupcanIntegration: false, integrations: {} })
+		).integrations as { [id: string]: Integration };
+		const senderId = sender.id ?? '';
+		if (!integrations.hasOwnProperty(senderId)) {
+			if (message?.action == 'register') {
+				//External extension wants to register
+				const reg_request = message as RegisterRequest;
+				integrations[senderId] = {
+					name: reg_request.name,
+					state: IntegrationStateDisabled,
+				};
+				api.storage.local.set({
+					integrations,
+					[EventKey]: {
+						type: MessageEvent,
+						message: `<p>The extension <b>${reg_request.name}</b> would like to integrate with BlueBlocker.<br>Visit the <a href="${api.runtime.getURL("/src/pages/integrations/index.html")}" target="_blank">integrations page</a> to complete set up.</p>`,
+						options: { html: true },
+					},
+				});
+				response = {
+					status: SuccessStatus,
+					result: 'integration registered',
+				} as SuccessResponse;
+				console.debug(
+					logstr,
+					refid,
+					`registered a new extention: ${reg_request.name} ${senderId}. ext resp:`,
+					response,
+				);
+				return;
+			}
+			response = { status: ErrorStatus, message: 'extension not allowed' } as ErrorResponse;
+			return;
+		}
```
I like this

rougetimelord commented on src/shared.ts:
```diff
@@ -645,16 +646,31 @@ export async function BlockBlueVerified(user: BlueBlockerUser, config: Config) {
+		if (disallowedWordsWithoutEmptyStrings.length > 0){
+			// this makes extra sure that emojis are always detected, regardless if they are attached to a word or another string of emojis.
+			// the 'i' makes the test case insensitive, which helps users not have to worry about typing the same word multiple times with different variations.
+			const disallowedWordsAndEmojis = new RegExp(`${config.disallowedWords.join('(?= |$)|')}(?= |$)`, 'i');
```
I'm noticing that in trying to make partial words not have false positives (eg. "maga" matching "magazine") I kind of mangled detecting emojis anywhere in the string, since there's now a positive lookahead for a space or EOL.

Uhhhh, I don't really know what the best way to handle emojis specially would be? Maybe:
```ts
const disallowedWordsAndEmojis = config.disallowedWords.reduce(
  (regex, value) => regex + value + (/*check if value is an emoji*/) ? '|' : '(?= |$)|',
  ''
)
```

rougetimelord replied:
pinging @MaxIsJoe, lmk if you have any suggestions here

MaxIsJoe replied:
We could run another check specifically for emojis, and we don't need to be specific on where they are placed or what pattern they're in.
you can find emojis by using the regex `/\p{Emoji}/u;`, filter all the emojis inside `config.disallowedWords` then run a simpler regex test along side `disallowedWordsAndEmojis` to see if any blacklisted emoji is mentioned anywhere.

kheina replied:
maybe
```js
const regexTokenRegex = /[\[\]\\\.\|\{\}\(\)\?\:\<\>\*\+\'\"\=\^\$]/gi; // this feels psychotic
let str = '[test]'
console.log(str.replace(regexTokenRegex, m => '\\' + m[0]))
\[test\]
```

rougetimelord replied:
my brain hurts

rougetimelord replied:
I think this follows through on all of your suggestions? https://github.com/kheina-com/Blue-Blocker/pull/267/commits/480ef4e16bc59f870e043b5852e5343331732f77

kheina commented:
I would like to do a full review before we merge anything this large, I'll do that tonight. I also have thoughts on what to do regarding the emoji/words regex which I'll add when I get to that part of the pr.

tonight being ~4-5 hours from now, I'm at work currently

rougetimelord commented on src/constants.ts:
```diff
@@ -7,9 +7,7 @@ try {
-			...browser.runtime,
-			OnInstalledReason: chrome.runtime.OnInstalledReason,
-			restartAfterDelay: chrome.runtime.restartAfterDelay,
+			...browser.runtime
```
> has this been tested on firefox and nothing has caused issues?

It shouldn't because we never use restartAfterDelay and chrome.runtime.OnInstalledReason should already be being  translated to browser.runtime.OnInstalledReason by Firefox according to [mdn](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Chrome_incompatibilities)

cooljeanius replied:
(this was #271 for reference)

rougetimelord commented on src/shared.ts:
```diff
@@ -645,16 +648,31 @@ export async function BlockBlueVerified(user: BlueBlockerUser, config: Config) {
+			// this makes extra sure that emojis are always detected, regardless if they are attached to a word or another string of emojis.
+			// the 'i' makes the test case insensitive, which helps users not have to worry about typing the same word multiple times with different variations.
+			const disallowedWordsAndEmojis = new RegExp(`${config.disallowedWords.join('(?= |$)|')}(?= |$)`, 'i');
+			const usernameToTest = (user.legacy.name).replace(/ {2,}/, '');
```
Grr oops

rougetimelord reviewed (commented):
ts is now giving me an error that api.runtime.sendMessage is not callable, due to incompatible signatures on the union type. It isn't a real problem but ts is being picky for some reason

kheina commented on src/popup/index.ts:
```diff
@@ -149,6 +158,33 @@ function exportSafelist() {
+	).filter(w => w);
+	api.storage.sync.set({ disallowedWords: {
+		list: wordList,
+		regExp: compileDisallowedWordsRegExp(wordList)
```
can you store regex in json??

rougetimelord replied:
no :(

kheina replied:
check out my pr

kheina commented on src/constants.ts:
```diff
@@ -77,14 +78,18 @@ export const ReasonNftAvatar: number = 1;
+	[ReasonDisallowedWordsOrEmojis]: 'disallowed words or emojis',
 };
 
+export const emojiRegExp = RegExp(/(\u00a9|\u00ae|[\u2000-\u3300]|\ud83c[\ud000-\udfff]|\ud83d[\ud000-\udfff]|\ud83e[\ud000-\udfff])/, 'g')
```
is this reliable? where did you find this?

rougetimelord replied:
https://stackoverflow.com/questions/60626009/regex-filter-special-characters-like-japanese-but-preserve-emojis

kheina replied:
that question specifies that that regex string captures more than just emoji characters, specifically japanese characters. the answer links to this repo: https://github.com/mathiasbynens/emoji-regex which we could use, or the one that I provided previously

rougetimelord replied:
The one you wrote up failed for emojis that are an emoji + zero width combiner + modifier :(

I also saw the emoji-regex package, my one concern is that the pattern by itself is 12.8 kB which uhhhh, suckssssss :(

kheina replied:
`[\p{Emoji_Presentation}\u200d]+/gu`? which emojis were you testing with? I was using the family ones

https://regex101.com/r/E4rmCw/1

rougetimelord replied:
oh huh, weird now it works, ok I'll switch it over

rougetimelord replied:
I was using 🏋️‍♂️, it still seems like it's only matching the zero-width and not the underlying weight lift emoji? But that's still better than matching Japanese

kheina commented:
this is what I meant by having a "compiled" config for running on vs saving. this way we're not saving extra data and only compiling once per request https://github.com/kheina-com/Blue-Blocker/pull/283

rougetimelord commented on src/shared.ts:
```diff
@@ -78,7 +81,7 @@ function unblockUser(
 	let url: string = '';
 
 	if (root.includes('tweetdeck')) {
-		url = 'https://api.twitter.com/1.1/';
+		url = 'https://api.x.com/1.1/';
```
The web client is reaching out to `x.com/i/api/2` now, unsure about what tweetdeck is doing nowadays but I don't think we *need* to continue support for tweet deck lol.

![image](https://github.com/kheina-com/Blue-Blocker/assets/14047458/d3ac0df0-afdb-42e1-9cc5-b524a2dde194)

kheina replied:
I normally just block and mute a person normally and confirm that __those__ endpoints use the the api. though, wouldn't be surprised if they bumped it to 2

was tweetdeck eliminated?

rougetimelord replied:
![image](https://github.com/kheina-com/Blue-Blocker/assets/14047458/62becce5-76be-4c2b-b251-65309e66b981)

Still 1.1 for blocks and mutes. Tweetdeck got turned into a blue only feature lmao

rougetimelord reviewed (commented):
commenting so I can close my review, github is confusing me

rougetimelord commented:
Going to start testing!

rougetimelord commented:
Problems:
- Empty disallowed words list seems to spam block everything
- Disallowed words blocking logic does not respect the `pause collection` setting
- TypeError in popup script, line 47: cannot read property of undefined " join"

MaxIsJoe commented:
> Problems:
> 
> * [ ]  Empty disallowed words list seems to spam block everything

![image](https://github.com/kheina-com/Blue-Blocker/assets/34368774/b9216ffd-9528-452d-a6c8-d8845e9c47fa)

I added this check specifically to address this issue in my original PR.

rougetimelord commented:
> > Problems:
> > 
> > * [ ]   Empty disallowed words list seems to spam block everything
> 
> ![image](https://private-user-images.githubusercontent.com/34368774/332230574-b9216ffd-9528-452d-a6c8-d8845e9c47fa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYyNjA1NzEsIm5iZiI6MTcxNjI2MDI3MSwicGF0aCI6Ii8zNDM2ODc3NC8zMzIyMzA1NzQtYjkyMTZmZmQtOTUyOC00NTJkLWE2YzgtZDg4NDVlOWM0N2ZhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTIxVDAyNTc1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUzZTRhMWJhM2JkY2ZhMDYzZTE4MmQ2Yjk5M2NiNWU4YjE5Yjc2MTFlODVhMTQ4NDZjODgxZWI5NWJmZTBjYmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.MeTIt9Oo9N5ejwgJMUsr7GPbplMb7EQYEkqSP8DvnBw)
> 
> I added this check specifically to address this issue in my original PR.

Yea, but then we switched to precompiling the regex, so I'll futz around with it

rougetimelord commented:
Ok that's all the bugs I've found so far squashed. Going to keep testing for tonight to see if something pops up, but I think we're close to being able to merge.

kheina commented:
I'll do another review and some testing tonight after work. hopefully it can go out tonight

kheina commented:
> I'll do another review and some testing tonight after work. hopefully it can go out tonight

I'll probably also just push a bunch of styling changes/fixes since I'm pretty particular about the appearance of the popup menu

rougetimelord commented:
> > I'll do another review and some testing tonight after work. hopefully it can go out tonight
> 
> I'll probably also just push a bunch of styling changes/fixes since I'm pretty particular about the appearance of the popup menu

Go ham! I think having the setting be hidden behind a checkbox if not enabled would make a lot of sense in terms of not cluttering up the interface, kinda like what the skip threshold setting does

kheina commented:
> > > I'll do another review and some testing tonight after work. hopefully it can go out tonight
> > 
> > 
> > I'll probably also just push a bunch of styling changes/fixes since I'm pretty particular about the appearance of the popup menu
> 
> Go ham! I think having the setting be hidden behind a checkbox if not enabled would make a lot of sense in terms of not cluttering up the interface, kinda like what the skip threshold setting does

I went ham

rougetimelord commented on src/popup/index.ts:
```diff
@@ -322,42 +343,61 @@ document.addEventListener('DOMContentLoaded', () => {
+			document
+				.getElementsByName('soupcan-integration-option')
+				.forEach(e => (e.style.display = 'flex'));
+		})
```
We should probably use this if block to migrate soupcan to the new extension flow sometime soon, but ¯\_(ツ)_/¯ for right now

kheina replied:
I set up a new workflow for integrations that send the full user block as opposed to just the handle, so it doesn't work with the new integration system

rougetimelord replied:
Hmm, might be a little side quest for me to do a PR over there at some point

rougetimelord replied:
Side quest complete: https://github.com/bethylamine/soupcan/pull/140

rougetimelord commented:
One approval pls 👉🏻👈🏻

kheina reviewed (approved):
way too much shit in one PR tbh. lets try to keep them smaller from here on out (even for the releases
